src/jdk.compiler/share/classes/com/sun/tools/javac/comp/ReflectMethods.java

```diff
@@ -1492,360 +1492,226 @@ public void visitIf(JCTree.JCIf tree) {
-                        }
-                    } else {
-                        List<Body.Builder> clBodies = new ArrayList<>();
-                        for (JCTree.JCCaseLabel cl : c.labels) {
-                            ccl = (JCTree.JCConstantCaseLabel) cl;
-                            pushBody(ccl, FunctionType.functionType(JavaType.BOOLEAN));
-
-                            Value expr = toValue(ccl.expr);
-                            expr = convert(expr, tree.selector.type);
-                            final Value labelResult;
-                            if (tree.selector.type.isPrimitive()) {
-                                labelResult = append(CoreOp.eq(localTarget, expr));
-                            } else {
-                                labelResult = append(CoreOp.invoke(
-                                        MethodRef.method(Objects.class, "equals", boolean.class, Object.class, Object.class),
-                                        localTarget, expr));
-                            }
+            FunctionType actionType = FunctionType.VOID;
 
-                            append(CoreOp._yield(labelResult));
-                            clBodies.add(stack.body);
+            List<Body.Builder> bodies = visitSwitchStatAndExpr(tree, tree.selector, target, tree.cases, actionType,
+                    tree.patternSwitch && !tree.hasUnconditionalPattern);
 
-                            // Pop label
-                            popBody();
-                        }
+            result = append(ExtendedOp.switchStatement(target, bodies));
+        }
 
-                        localResult = append(ExtendedOp.conditionalOr(clBodies));
-                    }
+        private List<Body.Builder> visitSwitchStatAndExpr(JCTree tree, JCExpression selector, Value target,
+                                                          List<JCTree.JCCase> cases, FunctionType caseBodyType,
+                                                          boolean isDefaultCaseNeeded) {
 
-                    append(CoreOp._yield(localResult));
-                    bodies.add(stack.body);
+            List<Body.Builder> bodies = new ArrayList<>();
+            Body.Builder defaultLabel = null;
+            Body.Builder defaultBody = null;
 
-                    // Pop labels
-                    popBody();
-                } else if (headCl instanceof JCTree.JCDefaultCaseLabel) {
-                    // @@@ Do we need to model the default label body?
-                    pushBody(headCl, FunctionType.VOID);
+            for (JCTree.JCCase c : cases) {
 
-                    append(CoreOp._yield());
-                    defaultLabel = stack.body;
+                Body.Builder caseLabel = visitCaseLabel(tree, selector, target, c);
+                Body.Builder caseBody = visitCaseBody(c, caseBodyType);
 
-                    // Pop label
-                    popBody();
+                if (c.labels.head instanceof JCTree.JCDefaultCaseLabel) {
```

Comment:
I'm not sure this is optimal - can't we add all labels and bodies (even default ones) and then, after the loop, only add a synthetic default label if `isDefaultCaseNeeded` is set?

Reply:
We want to have the default body as the last one.

Reply:
This simplifies the lowering as example.